keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,812 @@
+   ```
+
+5. embedded JSON, further specified by the user using tags:
+   ```go
+   type Foo struct {
+     // +nullable
+     // +kubebuilder:type=object
+     Json JSON
+   }
+   ```
+   specified by:
+   ```yaml
+   type: object
+   properties:
+     json:
+       type: object
+       x-kubernetes-preserve-unknown-fields: true
+       nullable: true
+   ```
+   and published the same way as OpenAPI v3. For OpenAPI v2 we have to drop `type: object` and `nullable: true` as the latter does not exist in v2 and `type: object` is too restrictive (does not allow `null` as value).
+    
+   Dropping value validation does not change anything.
+
+6. string maps:
+   ```go
+   // +kubebuilder:additionalProperties:minimum=42
+   map[string]int
+   ```
+   specified by:
+   ```yaml
+   type: object
+   additionalProperties:
+     type: integer
+     minimum: 42
+   ```
+   and published as OpenAPI v3 and v2 the same way.
+   
+   After dropping the value validation, we get:
+   ```yaml
+   type: object
+   additionalProperties:
+     type: integer
+   ```
+7. IntOrString:
+   ```go
+   type Foo struct {
+     // +kubebuilder:pattern=abc
+     // +kubebuilder:anyOf=minimum=42,maximum=50;minimum=52,maximum=60
+     X *IntOrString
+   }
+   
+   type IntOrString struct {
+     Integer *int
+     String *string
+   } + marshallers
+   ```
+   specified by:
+   ```yaml
+   x-kubernetes-int-or-string: true
+   pattern: abc
+   anyOf:
+   - minimum: 42
+     maximum: 50
+   - minimum: 52
+     maximum: 60
+   ```
+   or alternatively by:
+   ```yaml
+   x-kubernetes-int-or-string: true
+   allOf:
+   - anyOf: // optionally provided and accepted
+     - type: integer
+     - type: string
+   - pattern: abc
+     anyOf:
+     - minimum: 42
+       maximum: 50
+     - minimum: 52
+       maximum: 60
+   ```
+   The latter is what is also published as OpenAPI v3, but as `x-kubernetes-int-or-string: true` in OpenAPI v2 (neither `anyOf`, nor `pattern` exists in v2).
+   
+   After dropping the value validation, we get:
+   ```yaml
+   x-kubernetes-int-or-string: true
+   ```   
+
+Note: the OpenAPI v2 output given above is the same we get from the [OpenAPI v2 filtering in Kubernetes 1.14](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion.go#L30): that filtering in the CRD OpenAPI v2 publishing pipeline supports the unrestricted range of the CRD OpenAPI v3 language and turns it into a weaker variant understandable by kubectl 1.13 and 1.14. Hence, our changes above have no influence on the OpenAPI v2 publishing other than passing through of our new `x-kubernetes-*` vendor extensions.
+
+Also note that structural schemas enforce types and properties outside of logical junctors. OpenAPI v2 publishing will preserve these and hence, structural schemas lead to a more complete OpenAPI v2 spec, client-side validation and `kubectl explain` output.
```

Comment:
+1

Reply:
@roycaihw ^^ is a great improvement for your publishing code.